feat(rerank): implement end-to-end Rerank API support

Silentely opened the pull request:
## Summary

本 PR 为 AxonHub 添加完整的 Rerank（文档重排序）API 支持，实现了从数据结构、Transformer 层、业务逻辑层到 API 层的端到端功能。

### 新增功能

#### 1. 数据结构层 (`internal/objects/`)
- `RerankRequest`: 支持 `model`、`query`、`documents`、`top_n` 等参数
- `RerankResponse`: 包含排序结果列表
- `RerankResult`: 包含 `index`、`relevance_score`、`document` 字段

#### 2. Transformer 层 (`internal/llm/transformer/`)
- 扩展 `Transformer` 接口，新增 `Rerank` 方法
- 支持自定义 `HttpClient` 参数（代理、超时、证书配置）
- 实现 `RerankError` 类型，保留上游 HTTP 状态码
- 支持 OpenAI 和 Azure 平台的 URL 构建

#### 3. 业务逻辑层 (`internal/server/biz/`)
- 多 Channel 负载均衡支持
- 基于系统配置的重试机制（4xx 错误不重试）
- 模型映射支持（`Channel.ChooseModel`）
- 完整的输入校验：
  - `top_n` 必须为正整数且不超过文档数量
  - 文档列表不能包含空字符串
- 集成请求日志系统（Request + RequestExecution 记录）

#### 4. API 层 (`internal/server/api/`)
- 新增 `POST /v1/rerank` 端点
- 透传上游 HTTP 状态码
- 标准化错误响应格式

#### 5. 前端 SDK (`frontend/src/lib/`)
- 更新 `api-client.ts` 支持 Rerank API 调用

### 关于用量日志

由于 Rerank API 的响应中**不包含 token 用量信息**（与 Chat/Embedding API 不同），因此未将其添加到 UsageLog。用户可通过**请求日志**功能查询 Rerank 的调用情况，包括：
- 请求时间、延迟
- 使用的渠道和模型
- 请求/响应详情
- 执行记录

### 测试覆盖

#### 单元测试 (`internal/llm/transformer/openai/rerank_test.go`)
| 测试用例 | 覆盖场景 |
|----------|----------|
| `TestRerank_Success` | 成功请求、响应解析 |
| `TestRerank_ValidationErrors` | nil 请求、空 model/query/documents、top_n 边界校验 |
| `TestRerank_HTTPError` | 400 错误状态码提取 |
| `TestRerank_ServerError` | 500 错误状态码提取 |
| `TestRerank_WithTopN` | top_n 参数传递验证 |
| `TestBuildRerankURL` | URL 构建（标准/Azure/不同路径格式） |

#### 数据结构测试 (`internal/objects/rerank_test.go`)
- JSON 序列化/反序列化测试

#### API 测试验证
```bash
# 正常请求
curl -X POST http://localhost:5173/v1/rerank \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer <api-key>" \
  -d '{
    "model": "Qwen/Qwen3-Reranker-4B",
    "query": "What is machine learning?",
    "documents": ["Machine learning is...", "The weather is...", "Deep learning..."],
    "top_n": 2
  }'

# 响应示例
{"results":[{"index":0,"relevance_score":0.996},{"index":2,"relevance_score":0.041}]}

# 输入校验测试
# top_n 超出范围 -> 400: "top_n (10) cannot exceed the number of documents (2)"
# 空文档字符串 -> 400: "document at index 1 is empty"
# top_n 为负数 -> 400: "top_n must be a positive integer"
```

### 文件变更

```
 frontend/src/lib/api-client.ts                 |  93 +++++--
 internal/llm/constants.go                      |   1 +
 internal/llm/transformer/interfaces.go         |  12 +
 internal/llm/transformer/openai/rerank.go      | 164 ++++++++++++
 internal/llm/transformer/openai/rerank_test.go | 306 +++++++++++++++++++++++
 internal/objects/rerank.go                     |  38 +++
 internal/objects/rerank_test.go                |  70 ++++++
 internal/server/api/fx_module.go               |   1 +
 internal/server/api/rerank.go                  |  56 +++++
 internal/server/biz/fx_module.go               |   1 +
 internal/server/biz/rerank.go                  | 333 +++++++++++++++++++++++++
 internal/server/routes.go                      |   2 +
 12 files changed, 1061 insertions(+), 16 deletions(-)
```

## Test Plan

- [x] 单元测试全部通过 (`go test ./...`)
- [x] 构建验证通过 (`make build-backend`)
- [x] API 功能测试通过（成功请求、错误处理）
- [x] 请求日志记录验证通过
- [x] 输入校验测试通过

```
{
  "model": "Qwen/Qwen3-Reranker-4B",
  "query": "What is machine learning?",
  "documents": [
    "Machine learning is a subset of artificial intelligence.",
    "The weather is sunny today.",
    "Deep learning uses neural networks."
  ],
  "top_n": 2
}

{
  "results": [
    {
      "index": 0,
      "relevance_score": 0.9963386058807373
    },
    {
      "index": 2,
      "relevance_score": 0.04179020971059799
    }
  ]
}
```

Silentely committed:
feat(rerank): add Rerank API support

- Add RerankRequest, RerankResponse, RerankResult data structures
- Extend Transformer interface with Rerank method
- Implement OpenAI-compatible Rerank transformer
- Add RerankService business logic layer
- Add Rerank HTTP handler and route (POST /v1/rerank)
- Update frontend API client with rerank method

Silentely committed:
feat(rerank): integrate request logging for Rerank API

- Add APIFormatOpenAIRerank constant for request format tracking
- Inject RequestService into RerankService for logging
- Create request record before API call with model and request body
- Update channel_id after channel selection
- Update request status on success/failure with response body
- Add latency tracking for performance monitoring

Silentely committed:
fix(rerank): enhance with load balancing, retry, model mapping and validation

- Add httpClient parameter to Transformer.Rerank for proxy/timeout support
- Implement multi-channel load balancing and retry mechanism
- Add model mapping via Channel.ChooseModel
- Preserve upstream HTTP status codes with RerankError type
- Add comprehensive input validation (top_n bounds, empty documents)
- Add unit tests for transformer layer

Silentely committed:
refactor(rerank): address code review feedback

Changes based on Gemini Code Assist review:

1. Translate Chinese comment to English in rerank.go
2. Simplify URL building logic using strings.HasSuffix
3. Extract intPtr helper to shared xptr package to follow DRY principle
4. Remove unused RerankError type in biz/rerank.go
5. Rename mustMarshalJSON to marshalJSONWithFallback with error logging

looplj commented:
e2e 的意思是透传而不做转换吗？

looplj commented on internal/llm/rerank.go:
```diff
@@ -0,0 +1,38 @@
+package objects
+
+// RerankRequest represents a rerank request.
+type RerankRequest struct {
```
需要注意一下分层。
llm 是作为一个底层 package 的，objects 是业务层的，不能反向依赖。

looplj commented on internal/pkg/xptr/ptr.go:
```diff
@@ -0,0 +1,17 @@
+
+// IntPtr returns a pointer to the given int value.
+// Useful for creating pointers to int literals in test cases.
+func IntPtr(i int) *int {
```
这些东西应该都不需要，用 lo.ToPtr 即可。

looplj commented on internal/server/biz/rerank.go:
```diff
@@ -0,0 +1,325 @@
+	SystemService  *SystemService
+}
+
+func NewRerankService(params RerankServiceParams) *RerankService {
```
应该都需要复用 chat 那一套。
就算是透传，也应该在 chat 那一套设计一套方式。

Silentely committed:
refactor(rerank): address code review feedback - layer dependency fix

Changes based on maintainer review:

1. Remove xptr package, use lo.ToPtr instead (DRY principle)
2. Move RerankRequest/RerankResponse/RerankResult from objects to llm package
   - This fixes the layer dependency issue where llm (base layer) was importing
     objects (business layer)
3. Update Transformer interface to use llm types instead of objects types
4. Remove unused boolPtr helper, use lo.ToPtr instead
5. Add RerankUsage type in llm package for rerank-specific usage tracking

Note: The suggestion to "reuse chat pipeline design" requires further
architectural discussion with maintainer to determine the best approach.

Silentely committed:
refactor(rerank): align with Chat pipeline design pattern

- Add RerankChannelSelector interface with decorator pattern support
- Add RerankLoadBalancer interface for channel sorting
- Implement DefaultRerankChannelSelector and LoadBalancedRerankChannelSelector
- Use dependency injection with optional ChannelSelector parameter
- Fix UpdateRequestChannelID to use detached context
- Record mapped request body instead of original for execution record
- Remove unused channelID parameter from updateRequestSuccess

Silentely commented:
## 提交 aa2b7fe6: 对齐 Rerank 与 Chat Pipeline 设计模式

### 架构变更

**新增接口：**
- `RerankChannelSelector` - 支持装饰器模式的 Channel 选择接口
- `RerankLoadBalancer` - Channel 优先级排序接口

**新增实现：**
- `DefaultRerankChannelSelector` - 基础选择器，返回所有支持该模型的已启用 Channel
- `LoadBalancedRerankChannelSelector` - 装饰器实现，对 Channel 应用负载均衡排序

### Bug 修复

1. **Context 分离** - `UpdateRequestChannelID` 现使用 `DetachWithTimeout`，确保即使客户端断开连接，数据库写入仍能完成

2. **请求体一致性** - Execution 记录现存储映射后的请求（包含实际发送给 Channel 的模型名称），而非原始请求

3. **移除未使用参数** - 从 `updateRequestSuccess` 函数中移除未使用的 `channelID` 参数

### 设计对齐

接口设计现与 `internal/server/chat/channels.go` 中的 Chat Pipeline 模式对齐：
- `Select(ctx, req *llm.RerankRequest)` 对应 Chat 的 `Select(ctx, req *llm.Request)`
- 装饰器模式支持未来扩展（SelectedChannels 过滤器、Tags 过滤器等）
- 通过 `fx.In` 配合 `optional:"true"` 标签实现依赖注入

### 测试验证

✅ 构建通过
✅ 单元测试通过
✅ 使用 `Qwen/Qwen3-Reranker-4B` 模型的 API 集成测试通过

looplj commented:
想确认下，rerank 是不是可以和 embedding 一样，直接复用基于 chat 那一套基建，或者说是否有限制导致不能复用。
做两套后续可能难以维护。

Silentely commented:
> 想确认下，rerank 是不是可以和 embedding 一样，直接复用基于 chat 那一套基建，或者说是否有限制导致不能复用。 做两套后续可能难以维护。

Rerank不太适合直接用Embedding那种，格式不兼容那一套(Query和Documents是Rerank特有)，可以按照独立实现 + 抽取公共组件，或者这个可以先挂着后期作者看看有没有更好的实现方式

looplj commented:
代码我先合并，发个新版本，你可以用上。

我这边下个版本会进行一定得重构，将 llm request 和 Response 模型整理一下，更加通用，可以覆盖更多场景，一切都基于 llm 和 transformer，这个是架构基础。。
